code/datums/beam.dm

```diff
@@ -26,53 +27,76 @@
 	icon_state = beam_icon_state
 	beam_type = btype
 
+
 /datum/beam/proc/Start()
 	Draw()
-	while(!finished && target && world.time<endtime && get_dist(origin,target)<max_distance && origin.z == target.z)
+	while(!finished && origin && target && world.time < endtime && get_dist(origin,target)<max_distance && origin.z == target.z)
 		if(origin.loc != origin_oldloc || target.loc != target_oldloc)
+			origin_oldloc = origin.loc //so we don't keep checking against their initial positions, leading to endless Reset()+Draw() calls
+			target_oldloc = target.loc
 			Reset()
 			Draw()
 		sleep(sleep_time)
+
 	qdel(src)
 
+
 /datum/beam/proc/End()
 	finished = 1
 
+
 /datum/beam/proc/Reset()
 	for(var/obj/effect/ebeam/B in elements)
 		qdel(B)
 
+
 /datum/beam/Destroy()
 	Reset()
 	target = null
 	origin = null
 	return ..()
 
+
 /datum/beam/proc/Draw()
-	var/Angle=round(Get_Angle(origin,target))
+	var/Angle = round(Get_Angle(origin,target))
 
 	var/matrix/rot_matrix = matrix()
 	rot_matrix.Turn(Angle)
 
-	var/DX=(32*target.x+target.pixel_x)-(32*origin.x+origin.pixel_x)
-	var/DY=(32*target.y+target.pixel_y)-(32*origin.y+origin.pixel_y)
-	var/N=0
-	var/length=round(sqrt((DX)**2+(DY)**2))
-	
-	for(N,N<length,N+=32)
-		var/obj/effect/ebeam/X= new beam_type(origin.loc)
-		X.owner=src
+	//Translation vector for origin and target
+	var/DX = (32*target.x+target.pixel_x)-(32*origin.x+origin.pixel_x)
```

Comment:
ah yea I missed the "translation vectors" thing, lol. makes perfect sense in context.
might be helpful to clarify that it's in pixels but not really worth the effort if you won't be changing anything else